src/csharp/Grpc.Core/Internal/AsyncCall.cs

```diff
@@ -626,6 +626,14 @@ private void HandleFinished(bool success, ClientSideStatus receivedStatus)
             if (status.StatusCode != StatusCode.OK)
             {
                 streamingResponseCallFinishedTcs.SetException(new RpcException(status, receivedStatus.Trailers));
+                if (status.StatusCode == StatusCode.Cancelled)
```

Comment:
Even if we escape the using statement and cancel the call (due to the `grpc_call_unref` IIC), I'm not certain the status is guaranteed to be cancelled (in case there was e.g. a race with a "deadline exceeded" status).

Can we safely do the following Exception suppression, unconditionally?

Reply:
I didn't want to assume too much. In general users should be exhausting the streams, not doing so can lead to other problems (so it's fair that they are warned about an unobserved exception).
The case when the call is cancelled is one situation when not exhausting the response stream is safe, so I wanted to suppress in this special case.
I think one more thing I can check for is the "cancellationRequested" flag (whether Cancel() was called), which should make this supression even more reliable.

Reply:
Can we make that check, for if cancellation has been requested (or another equivalent check that indicates e.g. if `Dispose` has been called)?

Without that, my concern is that the issue described in https://github.com/grpc/grpc/issues/17458#issuecomment-552844393 can still happen, just more rarely (e.g. if the call has a deadline and it just happens to go off right around the time that we `Dispose` the call, causing the same type of problem but changing status to "deadline exceeded").

Reply:
Ah, I forgot about it. Done.

Reply:
The new check here of, 

`if (status.StatusCode == StatusCode.Cancelled || origCancelRequested)`

eliminates the problem for the case where the call was explicitly cancelled. But AFAICS, if we exit the using and `Dispose` the call, `cancelRequested` won't be triggered, right?

Can we also check if the call has been disposed (and i.e. "cancelled" due to the `grpc_call_unref`)?

Reply:
To clarify:
`call.Dispose()` for `Async*Call` objects exposed publicly actually calls "Cancel" under the hood, which should address your concern.

https://github.com/grpc/grpc/blob/5817f6287d758ead304668ea6e7819fd9a4c3dcb/src/csharp/Grpc.Core.Api/AsyncServerStreamingCall.cs#L109
And as you can see, the dispose action is `asyncCall.Cancel`.
https://github.com/grpc/grpc/blob/5817f6287d758ead304668ea6e7819fd9a4c3dcb/src/csharp/Grpc.Core/Calls.cs#L119